packages/components/src/Interactive.ts

```diff
@@ -35,124 +35,124 @@ const allowedHoverVariables = [
   'boxShadow',
   'opacity',
   'textDecoration',
-];
+] as const;
+
+// Union type of all values in allowedHoverVariables.
+type HoverStyleProperty = (typeof allowedHoverVariables)[number];
+
 const pointerTags = ['button', 'area', 'a', 'select'];
 const pointerTypes = ['radio', 'checkbox', 'submit', 'button'];
 
-export type InteractiveProps = {
-  children?: ReactNode;
-  className?: string;
-  disabled?: boolean;
-  hover?: React.CSSProperties;
-  role?: string;
-  style?: React.CSSProperties;
-  tabIndex?: number;
+// Since Interactive may render any HTML element based on the tagName prop,
+// use AllHTMLAttributes<HTMLElement> to allow this to accept attributes
+// for any HTML element.
+// TODO: Can a more specific type be used by parameterizing this with a tag name?
+export interface InteractiveProps extends AllHTMLAttributes<HTMLElement> {
+  /** Styles applied when element is hovered or focused. */
+  hover?: Pick<React.CSSProperties, HoverStyleProperty>;
+
+  /** HTML tag to render. */
   tagName?: keyof JSX.IntrinsicElements;
-  type?: string;
-  onBlur?: (event: any) => void;
-  onClick?: (event: any) => void;
-  onKeyDown?: (event: any) => void;
-  onMouseDown?: (event: any) => void;
-};
-
-export const Interactive: React.ForwardRefExoticComponent<InteractiveProps> = forwardRef(
-  (
+
+  // Allow arbitrary data attributes on the rendered element.
+  [dataAttribute: `data-${string}`]: string;
+}
+
+export const Interactive = forwardRef((props: InteractiveProps, ref: ForwardedRef<HTMLElement>) => {
+  const {
+    children,
+    className = '',
+    disabled,
+    hover = {},
+    role,
+    style = {},
+    tabIndex,
+    tagName: TagName = 'div',
+    type,
+    onBlur,
+    onClick,
+    onKeyDown,
+    onMouseDown,
+    ...otherProps
+  } = props;
+
+  const [outline, setOutline] = useState<string>();
+  const { cursor } = style;
+
+  const computedCursor = Utils.cond(
+    [!!cursor, () => cursor],
+    [disabled, () => 'not-allowed'],
+    [!!onClick || pointerTags.includes(TagName) || pointerTypes.includes(type!), () => 'pointer']
+  );
+
+  const computedTabIndex = Utils.cond(
+    [_.isNumber(tabIndex), () => tabIndex],
+    [disabled, () => -1],
+    [!!onClick, () => 0],
+    () => undefined
+  );
+
+  const computedRole = Utils.cond(
+    [!!role, () => role],
+    [onClick && !['input', ...pointerTags].includes(TagName), () => 'button'],
+    () => undefined
+  );
+
+  const cssVariables = _.flow(
```

Comment:
can we have a comment here on what the heck this is doing/asserting?

Reply:
Yes. Generally, I was trying to avoid changes to the body of the Interactive function in this PR.

I think that with the stricter type on the `hover` prop, we could probably remove the assertion here since TS should prevent that case.

Reply:
also, I wonder if this could maybe be achieved with Typescript enforcement of non-optional props to the component?

Reply:
Yeah, in TypeScript, the reason for this assert is removed by https://github.com/DataBiosphere/terra-ui/pull/4255/commits/8f84f57abaaf53881109eba5be690d5f7a59ffd7. This would warn if you tried to do something like change padding via `hover`. It would warn that the CSS powering `hover` doesn't support that property. With this PR, the `hover` prop now only accepts style properties that are supported by the CSS.

Reply:
Updated in https://github.com/DataBiosphere/terra-ui/pull/4255/commits/ec197514bc70bc42ba92eb92505dba565b50d94e.

Reply:
nice